[CI] Failure in org.elasticsearch.snapshots.SnapshotStatusApisIT.testCorrectCountsForDoneShards

This failed 3 times over the last month on `master`.

```
./gradlew ':server:internalClusterTest' --tests "org.elasticsearch.snapshots.SnapshotStatusApisIT.testCorrectCountsForDoneShards" -Dtests.seed=B18A21A71F791680 -Dtests.security.manager=true -Dtests.locale=ar-MA -Dtests.timezone=Europe/Prague -Druntime.java=11
```

fails as:

```
  2> REPRODUCE WITH: ./gradlew ':server:internalClusterTest' --tests "org.elasticsearch.snapshots.SnapshotStatusApisIT.testCorrectCountsForDoneShards" -Dtests.seed=B18A21A71F791680 -Dtests.security.manager=true -Dtests.locale=ar-MA -Dtests.timezone=Europe/Prague -Druntime.java=11
  2> java.lang.AssertionError: 
    Expected: is <DONE>
         but: was <FAILURE>
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.junit.Assert.assertThat(Assert.java:956)
        at org.junit.Assert.assertThat(Assert.java:923)
        at org.elasticsearch.snapshots.SnapshotStatusApisIT.testCorrectCountsForDoneShards(SnapshotStatusApisIT.java:231)
```

example failure: https://gradle-enterprise.elastic.co/s/fypm6bjxyhz5m/console-log/raw 

=> looking into it

--- a/server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/status/SnapshotIndexShardStage.java
+++ b/server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/status/SnapshotIndexShardStage.java
@@ -31,7 +31,7 @@ public enum SnapshotIndexShardStage {
      */
     STARTED((byte)1, false),
     /**
-     * Snapshot metadata is being written
+     * Snapshot metadata is being written or this shard's status in the cluster state is being updated
      */
     FINALIZE((byte)2, false),
     /**

--- a/server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/status/TransportSnapshotsStatusAction.java
+++ b/server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/status/TransportSnapshotsStatusAction.java
@@ -176,6 +176,16 @@ public class TransportSnapshotsStatusAction extends TransportMasterNodeAction<Sn
                                 SnapshotIndexShardStatus shardStatus = shardStatues.get(shardEntry.key);
                                 if (shardStatus != null) {
                                     // We have full information about this shard
+                                    if (shardStatus.getStage() == SnapshotIndexShardStage.DONE
+                                            && shardEntry.value.state() != SnapshotsInProgress.ShardState.SUCCESS) {
+                                        // Unlikely edge case:
+                                        // Data node has finished snapshotting the shard but the cluster state has not yet been updated
+                                        // to reflect this. We adjust the status to show up as snapshot metadata being written because
+                                        // technically if the data node failed before successfully reporting DONE state to master, then
+                                        // this shards state would jump to a failed state.
+                                        shardStatus = new SnapshotIndexShardStatus(shardEntry.key, SnapshotIndexShardStage.FINALIZE,
+                                                shardStatus.getStats(), shardStatus.getNodeId(), shardStatus.getFailure());
+                                    }
                                     shardStatusBuilder.add(shardStatus);
                                     continue;
                                 }

